v2/handler.js

```diff
@@ -0,0 +1,241 @@
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+'use strict';
+
+var TypedError = require('error/typed');
+var EventEmitter = require('events').EventEmitter;
+var util = require('util');
+
+var v2 = require('./index');
+
+module.exports = TChannelV2Handler;
+
+var TChannelUnhandledFrameTypeError = TypedError({
+    type: 'tchannel.unhandled-frame-type',
+    message: 'unhandled frame type {typeCode}',
+    typeCode: null
+});
+
+var TChannelApplicationError = TypedError({
+    type: 'tchannel.application',
+    message: 'tchannel application error code {code}',
+    code: null,
+    arg1: null,
+    arg2: null,
+    arg3: null
+});
+
+function TChannelV2Handler(channel, options) {
+    if (!(this instanceof TChannelV2Handler)) {
+        return new TChannelV2Handler(channel, options);
+    }
+    EventEmitter.call(this);
+    var self = this;
+    self.channel = channel;
+    self.writeFrame = options.writeFrame;
+    // TODO: may be better suited to pull out an operation collection
+    // abstraction and then encapsulate through that rather than this
+    // run/complete approach
+    self.runInOp = options.runInOp;
+    self.completeOutOp = options.completeOutOp;
+    self.remoteHostPort = null; // filled in by identify message
+    self.lastSentFrameId = 0;
+}
+
+util.inherits(TChannelV2Handler, EventEmitter);
+
+TChannelV2Handler.prototype.nextFrameId = function nextFrameId() {
+    var self = this;
+    self.lastSentFrameId = (self.lastSentFrameId + 1) % v2.Frame.MaxId;
+    return self.lastSentFrameId;
+};
+
+TChannelV2Handler.prototype.handleFrame = function handleFrame(frame) {
+    var self = this;
+    switch (frame.body.type) {
+        case v2.Types.InitRequest:
+            return self.handleInitRequest(frame);
+        case v2.Types.InitResponse:
+            return self.handleInitResponse(frame);
+        case v2.Types.CallRequest:
+            return self.handleCallRequest(frame);
+        case v2.Types.CallResponse:
+            return self.handleCallResponse(frame);
+        case v2.Types.Error:
+            return self.handleError(frame);
+        default:
+            self.emit('error', TChannelUnhandledFrameTypeError({
+                typeCode: frame.body.type
+            }));
+    }
+};
+
+TChannelV2Handler.prototype.handleInitRequest = function handleInitRequest(reqFrame) {
+    var self = this;
+    if (self.remoteHostPort !== null) {
+        self.emit('error', new Error('duplicate init request')); // TODO typed error
+        return;
+    }
+    var hostPort = reqFrame.body.hostPort;
+    var processName = reqFrame.body.processName;
+    self.remoteHostPort = hostPort;
+    // TODO: use processName
+    self.emit('identify.in', hostPort, processName);
```

Comment:
these names were already chosen when we factored out v1/handler, changing them could happen in future